scripts/utils.py

```diff
@@ -337,3 +344,403 @@ def add_population_to_dataframe(
+    * Generating a series that underestimates the real one, because of treating missing values as zeros.
+
+    To avoid these problems, we first define a list of "big countries" that must be present in the data, in order to
+    safely do the aggregation. If any of these countries is not present for a particular variable and year, the
+    aggregation will be nan for that variable and year. Otherwise, if all big countries are present, any other missing
+    country will be assumed to have zero contribution to the variable.
+    For example, when aggregating the electricity demand of North America, United States and Mexico cannot be missing,
+    because otherwise the aggregation would significantly underestimate the true electricity demand of North America.
+
+    Additionally, the aggregation of a particular variable for a particular year cannot have too many nans. If the
+    number of nans exceeds num_allowed_nans_per_year, or if the fraction of nans exceeds frac_allowed_nans_per_year, the
+    aggregation for that variable and year will be nan.
+
+    Parameters
+    ----------
+    df : pd.Dataframe
+        Original dataset, which may contain data for that region (in which case, it will be replaced by the ).
+    region : str
+        Region to add.
+    min_frac_individual_population : float
+        Minimum fraction of the total population of the region that each of the big countries must exceed.
+    min_frac_cumulative_population : float
+        Minimum fraction of the total population of the region that the sum of the big countries must exceed.
+    reference_year_to_choose_countries_that_must_be_present : int
+        Reference year used to choose big countries.
+    num_allowed_nans_per_year : int or None
+        Maximum number of nans that can be present in a particular variable and year. If exceeded, the aggregation will
+        be nan.
+    frac_allowed_nans_per_year : float or None
+        Maximum fraction of nans that can be present in a particular variable and year. If exceeded, the aggregation
+        will be nan.
+    country_col : str
+        Name of country column.
+    year_col : str
+        Name of year column.
+    aggregations : dict or None
+        Aggregations to execute for each variable. If None, the contribution to each variable from each country in the
+        region will be summed. Otherwise, only the variables indicated in the dictionary will be affected. All remaining
+        variables will be nan.
+    countries_regions : pd.DataFrame or None
+        Countries-regions dataset. If None, it will be loaded from owid catalog.
+    population : pd.DataFrame or None
+        Population dataset. If None, it will be loaded from owid catalog.
+
+    Returns
+    -------
+    df_updated : pd.DataFrame
+        Original dataset after adding (or replacing) data for selected region.
+
+    """
+    # TODO: Add tests.
```

Comment:
I'd personally appreciate to have a single "integration" test with real data rather than tons of unit tests with made up data. With an integration test I see what problem have you solved with it and it also makes it easy to put breakpoint in the function, run tests and immediately get a sense how does it work. And it's easier :).

Reply:
I think both should exist. Having some unit tests on made-up data makes it easier to test individual functionalities, or  fringe cases that we may not often see in real data. We can talk about this later.